book/src/asm/machines.md

```diff
@@ -2,45 +2,45 @@
 
 Machines are the first main concept in powdr asm. They can currently be of two types: dynamic or static.
 
-## Dynamic machines
+## Virtual machines
 
 Dynamic machines are defined by:
 - a degree, indicating the number of execution steps
-- a set of registers, including a program counter
-- an instruction set
+- a set of [registers](./registers.md), including a program counter
+- an [instruction set](./instructions.md)
 - constraints
-- a set of functions
+- a set of [functions](./functions.md)
 - a set of submachines
 
-> Dynamic machines are currently limited to having a single function called `main`
-
 An example of a simple dynamic machine is the following:
 
 ```
 {{#include ../../../test_data/asm/book/hello_world.asm}}
 ```
 
-## Static machines
+## Constrained machines
 
-Static machines are a lower-level type of machine. They do not have registers, and instead rely on simple committed and fixed columns. They are used to implement hand-optimized computation.
+Constrained machines are a lower-level type of machine. They do not have registers, and instead rely on simple committed and fixed columns. They are used to implement hand-optimized computation.
 
 They are defined by:
 - a degree, indicating the number of execution steps
-- a set of functions
-- a function identifier column, used to make constraints conditional over which function is called
-- a latch column, used to identify rows at which the machine can be accessed from the outside (where the inputs and outputs are passed)
+- a set of [operations](./operations.md)
+- an `operation_identifier` column, used to make constraints conditional over which function is called
+- a `latch` column, used to identify rows at which the machine can be accessed from the outside (where the inputs and outputs are passed)
+- a set of submachines
+- a set of [links](links.md)
 
-An example of a simple static machine is the following:
+An example of a simple constrained machine is the following:
 
 ```
 {{#include ../../../test_data/asm/book/simple_static.asm}}
 ```
 
-For more details on the constraints, check out the [pil](../pil) section of this book. Note that the parameters of the function are columns declared within the constraints block.
+For more details on the constraints, check out the [pil](../pil) section of this book. Note that the parameters of the operation are columns declared within the constraints block.
 
 ## Submachines
 
-Machines can have submachines which they access by defining [external instructions](./instructions.md). They are declared as follows:
+Machines can have submachines which they access by defining [external instructions](./instructions.md) or [links](./links.md). They are declared as follows:
```

Comment:
Wait, what's the difference between external instructions and links?

Reply:
Instructions are a virtual machine concept. Links are a constrained machine concept. Links rely on a user-defined boolean flag and user-defined columns. External instructions rely on an instruction name and assignment registers, and get compiled down to links where the link flag is the instruction flag (introduced internally). The inputs and outputs are the same as in the instruction declaration as assignment registers compile down to committed columns of the same name.